### PR TITLE
At4 758 implement cmake coding standards

### PR DESCRIPTION
**Pull request by mitchmickaliger**
Updated all CMake files to match new CMake documentation in cheetah repo.

### CODE REVIEW

File: cmake/doxygen.cmake [hunk 1]
```diff
@@ -1,50 +1,63 @@
...
+
+else()
     add_custom_target(doc ${doc_all_target}
-      COMMAND ${CMAKE_COMMAND} -E echo
-      COMMENT "No doc target configured. Rebuild with -DENABLE_DOC=true" VERBATIM
+                      COMMAND ${CMAKE_COMMAND} -E echo
+                      COMMENT "No doc target configured. Rebuild with -DENABLE_DOC=true" VERBATIM
     )
-endif(ENABLE_DOC)
+endif()
```
Comment (chrisjwilliams) [inline cmake/doxygen.cmake]: leave ENABLE_DOC in for clarity

File: src/cpp/CMakeLists.txt [hunk 2]
```diff
@@ -86,26 +90,26 @@ if (DEBUG_PRINT)
...
-    else(ENABLE_DOC)
+    else()
         message(STATUS "  Doc:       Disabled (-DENABLE_DOC=true to enable)")
-    endif(ENABLE_DOC)
+    endif()
     message(STATUS "  Other:     ${CMAKE_INSTALL_PREFIX}/${MODULES_INSTALL_DIR}")
     message(STATUS "****************************************************************************")
     message(STATUS "")
-endif (DEBUG_PRINT)
+endif()
```
Comment (chrisjwilliams) [inline src/cpp/CMakeLists.txt]: leave in DEBUG_PRINT for clarity

File: src/cpp/cmake/boost.cmake [hunk 3]
```diff
@@ -1,21 +1,35 @@
...
+if(NOT BOOST_GUARD_VAR)
+    set(BOOST_GUARD_VAR TRUE)
+else()
+    return()
+endif()
+
 # Check if the user wants to use the included boost, regardless of if they have a existing install
 if(NOT USE_THIRDPARTY_BOOST)
-    if(NOT Boost_FOUND)
+    if(NOT BOOST_FOUND)
```
Comment (chrisjwilliams) [inline src/cpp/cmake/boost.cmake]: Wont work. FindBoost sets Boost_FOUND, not BOOST_FOUND

File: src/cpp/cmake/boost.cmake [hunk 4]
```diff
@@ -1,21 +1,35 @@
...
         message(STATUS "Checking for an installed boost...")
-        find_package(Boost QUIET) # Check for an existing boost install. NB we fail quietly here, as we can default to using the included boost.
-        if(Boost_FOUND)
-            message(STATUS "Boost found (version ${Boost_MAJOR_VERSION}.${Boost_MINOR_VERSION})")
+        # Check for an existing boost install
+        find_package(Boost)
+        set(BOOST_FOUND ${Boost_FOUND})
+        if(BOOST_FOUND)
+            set(BOOST_INCLUDE_DIRS ${Boost_INCLUDE_DIRS})
```
Comment (chrisjwilliams) [inline src/cpp/cmake/boost.cmake]: you need to make this look like findBoost's variables, which differ from cheetah conventions.

File: src/cpp/cmake/boost.cmake [hunk 5]
```diff
@@ -1,21 +1,35 @@
...
+# If no boost install is found, use the included third-party boost/units
+if(NOT BOOST_FOUND)
     message(STATUS "Using thirdparty boost.")
-    # We do noy use find_package(Boost ...) here, as that calls FindBoost.cmake, which requires libraries to exist, and units is header-only
-    set(Boost_INCLUDE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/thirdparty/boost/1.65 CACHE PATH "" FORCE)
-    set(Boost_MAJOR_VERSION 1)
-    set(Boost_MINOR_VERSION 65)
+    # We do not use find_package(Boost ...) here, as that calls FindBoost.cmake, which requires libraries to exist, and boost::units is header-only
+    set(BOOST_INCLUDE_DIRS ${CMAKE_CURRENT_SOURCE_DIR}/thirdparty/boost/1.65 CACHE PATH "" FORCE)
+    set(BOOST_MAJOR_VERSION 1)
```
Comment (chrisjwilliams) [inline src/cpp/cmake/boost.cmake]: Also must set the vars that BoostFind uses

File: src/cpp/cmake/valgrind.cmake [hunk 6]
```diff
@@ -1,40 +1,57 @@
...
+                           COMMAND ;
+                           COMMENT "see ${outputdir}/${outputname} for the valgrind report."
         )
-        ADD_DEPENDENCIES(${valgrind_target} ${valgrind_target_local})
-    endmacro(VALGRIND_TARGET)
-endif(NOT VALGRIND_PATH) 
-
+        add_dependencies(${valgrind_target} ${valgrind_target_local})
+    endfunction(valgrind_target)
+endif()
```
Comment (chrisjwilliams) [inline src/cpp/cmake/valgrind.cmake]: pu back NOT VALGRIND_EXE for clarity

File: src/cpp/cmake/subpackage.cmake [hunk 7]
```diff
@@ -1,148 +1,193 @@
...
-    set(subpackage_path "${subpackage_parent}")
-    set(subpackage_namespace "${subpackage_parent}")
-    set(subpackage_top "")
+    # Restore subpackage tracking variables to the parent scope
+    set(SUBPACKAGE_PATH "${_subpackage_parent}")
+    set(_subpackage_namespace "${_subpackage_parent}")
 
-    set(SUBPACKAGE_WORK_PATH "${SUBPACKAGE_WORK_DIR}/${subpackage_path}")
-    set(SUBPACKAGE_FILENAME "${SUBPACKAGE_WORK_PATH}/${SUBPACKAGE_NAME}.cmake")
+    set(subpackage_work_path "${subpackage_work_dir}/${SUBPACKAGE_PATH}")
```
Comment (chrisjwilliams) [inline src/cpp/cmake/subpackage.cmake:151]: these two lines do nothing now this is a function. Are you sure the functionsality that this adds actually works? This functionality, I suspect, is to do with allowing multiple levels of subpacakge() calls, like a stack.

**Review by chrisjwilliams — CHANGES_REQUESTED** [review]

File: cmake/doxygen.cmake (hunk 1 above)
Comment (mitchmickaliger) [reply cmake/doxygen.cmake]: Fixed (also re-added to else())

File: src/cpp/CMakeLists.txt (hunk 2 above)
Comment (mitchmickaliger) [reply src/cpp/CMakeLists.txt]: Fixed

File: src/cpp/cmake/boost.cmake (hunk 3 above)
Comment (mitchmickaliger) [reply src/cpp/cmake/boost.cmake]: Fixed

File: src/cpp/cmake/valgrind.cmake (hunk 6 above)
Comment (mitchmickaliger) [reply src/cpp/cmake/valgrind.cmake]: Fixed

File: src/cpp/cmake/subpackage.cmake (hunk 7 above)
Comment (mitchmickaliger) [reply src/cpp/cmake/subpackage.cmake:151]: These are the two lines at the end of subpackage(), right? I've left that as a macro; include_subpackage_files() is also still a macro. Only subpackage_file() and test_utils() are functions in this file.

File: src/cpp/cmake/boost.cmake (hunk 4 above)
Comment (mitchmickaliger) [reply src/cpp/cmake/boost.cmake]: Now that I've added the cheetah CMake conventions doc to this repo, don't we want to follow those same conventions? Or do we really want to do things differently here?

File: src/cpp/cmake/boost.cmake (hunk 5 above)
Comment (mitchmickaliger) [reply src/cpp/cmake/boost.cmake]: See the comment in the issue directly above this one

File: src/cpp/cmake/boost.cmake (hunk 4 above)
Comment (chrisjwilliams) [reply src/cpp/cmake/boost.cmake]: we want to do both - other cmake dependecies might dpeend on the standard FindBoost conventions.

Comment (mitchmickaliger) [reply src/cpp/cmake/boost.cmake]: Ok, yeah, that makes sense. All of the variables output by find_package(Boost) are still present, so this should be fine.

File: src/cpp/cmake/boost.cmake (hunk 5 above)
Comment (mitchmickaliger) [reply src/cpp/cmake/boost.cmake]: Yes, I've changed the if() conditional here to if(NOT Boost_FOUND), and set these same variables again in native FindBoost.cmake format.

**Review by chrisjwilliams — APPROVED** [review]

**Pull request merged**